### ISSUE TITLE
Ref and Metadata should be resolved together in P2P case

### ISSUE THREAD

**Issue by mwleeds** [flatpak/flatpak#1592]
@alexlarsson said in https://github.com/flatpak/flatpak/pull/1586#issuecomment-382315160:

> The revision of the summary, metadata branch and actually installed commit can all differ, because we're not relying on a consistent snapshot of the entire repo (which the summary is in the non-p2p case). This means any decisions we make based on the metadata (say dependencies, or permissions like "app needs foo, is that ok") before we install can be invalid wrt to the app that we actually install.
> 
> For example, we might have an offline situation with twopeers, one with app A and one with app B where B is more up-to-date. Say we want app A, but the B peer has a newer metadata branch. I believe this means we will pull the metadata branch from B, correct? But the metadata in B talks about a newer version of A, and if A recently switched to a new runtime version we will be downloading the wrong (new) runtime for A, but when running the old A it will want the old runtime.

**Comment by mwleeds** [comment]
There's a relevant comment [here](https://github.com/flatpak/flatpak/blob/63052a51d66cccf4cad68c8f57b8fc5110ae1b15/common/flatpak-dir.c#L6455), but in fact this is worse in the P2P case, because  you can get in a bad situation just by having peers with different versions of `ostree-metadata`, without needing an inopportune update on the server as you do in the non-P2P case.

**Comment by mwleeds** [comment]
This is made even more complicated when you're dealing with more than one ref because each one can come from a different remote and there's no single "right" `ostree-metadata` that would correspond to the right versions of every ref. Maybe the solution is to fetch the commit metadata when we're trying to figure out dependencies and get the flatpak metadata from there? That still leaves the issue that the other information we will get from `ostree-metadata`, such as what refs are available, may not match what other peers have. It seems like we need to somehow consider outdated `ostree-metadata` refs, not just the most recent one available (https://github.com/flatpak/flatpak/issues/1473).

**Comment by mwleeds** [comment]
Looks like there's overlap between this and #1583. That one is also discussing mismatches between the metadata in `ostree-metadata` and the commit.

**Comment by alexlarsson** [comment]
So, say a peer has some apps A and B, and an ostree-metadata branch. Irrespective of other peers, are there any guarantees that the versions of A and B match in time both wrt each other and the ostree metadata?

I wonder if for install/updates we should not instead download the actual commit object and extract the metadata  from that instead. As long as we pull the exact commit checksum we took the metadata from we should be avoiding issues with the metadata not matching the version.

Of course, there is a reason we put the metadata in the summary/branch. If you're downloading and displaying information for all refs in a remote for e.g. display or searching, then downloading each commit is way to slow, so there is still an issue with this. It may be less problematic though. For instance, an app that asks the user if the permissions are ok before he clicks install could pass in the chosen permissions into the install and it could verify that the actual app doesn't request more permissions.

**Comment by pwithnall** [comment]
> So, say a peer has some apps A and B, and an ostree-metadata branch. Irrespective of other peers, are there any guarantees that the versions of A and B match in time both wrt each other and the ostree metadata?

Depends on how the server is updated. If we’re just talking about race conditions, it should be possible to atomically update the `A`, `B` and `ostree-metadata` refs by adding new commits for them, then updating the `summary` file with the new checksums for all three refs at once.

> I wonder if for install/updates we should not instead download the actual commit object and extract the metadata from that instead. As long as we pull the exact commit checksum we took the metadata from we should be avoiding issues with the metadata not matching the version.

Yes, that is exactly what I was going to suggest: downloading the `.commit` object (i.e. just the metadata for the commit) and extracting from that. That’s the only way I can think of to get a strong binding between the app metadata and the app binary. It feels right as an approach.

> Of course, there is a reason we put the metadata in the summary/branch. If you're downloading and displaying information for all refs in a remote for e.g. display or searching, then downloading each commit is way to slow, so there is still an issue with this. It may be less problematic though. For instance, an app that asks the user if the permissions are ok before he clicks install could pass in the chosen permissions into the install and it could verify that the actual app doesn't request more permissions.

You could put the unsigned cache of app metadata in the `summary` (and not in `ostree-metadata`) and then peers could update the cache to suit the set (and versions) of apps they are offering to their peers. As everything in the `summary` is unsigned, you would have to verify the permissions (and other metadata) against the signed commit metadata once you’ve downloaded the commit, but an unsigned cache is probably good enough for listing or searching things in gnome-software. Regenerating the cache on each peer also means that they only advertise the apps they have available, which would reduce the metadata size. Storing the app metadata in `ostree-metadata` would mean that all peers have to serve the metadata for all apps available on (say) flathub, which could get big.

In summary, my recommendations would be:
 * Put app metadata in the signed commit metadata for the app’s commit
 * Don’t put app metadata in `ostree-metadata`
 * Allow app metadata to be cached in unsigned `summary` files
 ** Use that for listing/searching in gnome-software
 ** Verify it matches the signed app metadata in the commit metadata when actually installing/updating an app
 ** The apps in this metadata cache can be tailored to each peer

**Comment by alexlarsson** [comment]
This is fixed by https://github.com/flatpak/flatpak/pull/1796
